plugins/bcm57xx/fu-bcm57xx-recovery-device.c

```diff
@@ -0,0 +1,753 @@
+	return fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+						     REG_NVM_SOFTWARE_ARBITRATION, tmp.r32,
+						     error);
+}
+
+static gboolean
+fu_bcm57xx_recovery_device_nvram_wait_done (FuBcm57xxRecoveryDevice *self, GError **error)
+{
+	BcmRegNVMCommand tmp = { 0 };
+	g_autoptr(GTimer) timer = g_timer_new ();
+	do {
+		if (!fu_bcm57xx_recovery_device_bar_read (self,
+							  FU_BCM57XX_BAR_DEVICE,
+							  REG_NVM_COMMAND, &tmp.r32,
+							  error))
+			return FALSE;
+		if (tmp.bits.Done)
+			return TRUE;
+		if (g_timer_elapsed (timer, NULL) > 0.2)
+			break;
+	} while (TRUE);
+
+	/* timed out */
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_TIMED_OUT,
+			     "timed out");
+	return FALSE;
+}
+
+static gboolean
+fu_bcm57xx_recovery_device_nvram_clear_done (FuBcm57xxRecoveryDevice *self, GError **error)
+{
+	BcmRegNVMCommand tmp = { 0 };
+	tmp.bits.Done = 1;
+	return fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+						 REG_NVM_COMMAND, tmp.r32, error);
+}
+
+static gboolean
+fu_bcm57xx_recovery_device_nvram_read (FuBcm57xxRecoveryDevice *self,
+				       guint32 address, guint32 *buf, gsize bufsz,
+				       GError **error)
+{
+	for (guint i = 0; i < bufsz; i++) {
+		BcmRegNVMCommand tmp = { 0 };
+		guint32 val32 = 0;
+		if (!fu_bcm57xx_recovery_device_nvram_clear_done (self, error))
+			return FALSE;
+		if (!fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+						       REG_NVM_ADDR, address, error))
+			return FALSE;
+		tmp.bits.Doit = 1;
+		/* write one word at a time to avoid data corruption issue */
+		tmp.bits.First = TRUE;
+		tmp.bits.Last = TRUE;
+		if (!fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+							   REG_NVM_COMMAND, tmp.r32, error))
+			return FALSE;
+		if (!fu_bcm57xx_recovery_device_nvram_wait_done (self, error)) {
+			g_prefix_error (error, "failed to read @0x%x: ", address);
+			return FALSE;
+		}
+		if (!fu_bcm57xx_recovery_device_bar_read (self, FU_BCM57XX_BAR_DEVICE,
+							  REG_NVM_READ, &val32, error))
+			return FALSE;
+		buf[i] = GUINT32_FROM_BE(val32);
+		address += sizeof(guint32);
+		fu_device_set_progress_full (FU_DEVICE (self), i, bufsz);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_bcm57xx_recovery_device_nvram_write (FuBcm57xxRecoveryDevice *self,
+					guint32 address, const guint32 *buf, gsize bufsz,
+					GError **error)
+{
+	for (guint i = 0; i < bufsz; i++) {
+		BcmRegNVMCommand tmp = { 0 };
+		if (!fu_bcm57xx_recovery_device_nvram_clear_done (self, error))
+			return FALSE;
+		if (!fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+							   REG_NVM_WRITE, GUINT32_TO_BE(buf[i]), error))
+			return FALSE;
+		if (!fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+							   REG_NVM_ADDR, address, error))
+			return FALSE;
+		tmp.bits.Wr = 1;
+		tmp.bits.Doit = 1;
+		tmp.bits.First = 1; /* this is really slow but prevents corruption */
```

Comment:
I've updated the code in my repo to properly bust write: https://github.com/meklort/bcm5719-fw/commit/85e0e3fe2ba84b54fb0282fdfae1775e6e25e6cf

Effectively, write must not cross a page boundary. Once .First and .Last are set based on the page, everything is working correctly with the burst writes.

Reply:
I think for fwupd, where we're only using the recovery device to flash a **complete** backup image we can simplify things quite a bit and just check the address is aligned with 256, and the bufsz_dwrds is modulo 64.

Then I think this might just work:

		tmp.bits.First = i % page_size_dwrds == 0;
		tmp.bits.Last = (i + 1) % page_size_dwrds == 0;

I'll test in a minute on actual hardware. Thanks for figuring it out and explaining it to me!